### PR TITLE
[CIT-335] Device monitoring with collectd

### PR DESCRIPTION
**Pull request by albinsuresh**
## Proposed changes

Enable device monitoring with collectd by reading stats published by collectd and converting those stats to Thin Edge JSON measurements published back to the local MQTT bus

## Types of changes

What types of changes does your code introduce to `thin-edge.io`?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
https://cumulocity.atlassian.net/browse/CIT-335

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [x] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

## Further comments

* Stats published by collectd are read from the `collectd/` topic
* The measurement keys are derived from the topic name
* The measurement value is derived from the collectd MQTT payload
* The stats published by collectd within a 1000ms window are batched together into a single Thin Edge JSON message
* If the same stat value is repeated within this window, the last value is used
* The timestamp in individual collectd stat messages are ignored
* The timestamp for an entire batch is generated at the time of the receipt of the first message
* The Thin Edge JSON message created from a batch of collect stats is published back to `tedge/measurements` topic

### CODE REVIEW

File: dm-agent/src/main.rs [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+#[instrument(name = "main")]
+async fn run() -> Result<(), anyhow::Error> {
+    let config = mqtt_client::Config::new(DEFAULT_HOST, DEFAULT_PORT);
+    let mqtt = Arc::new(Client::connect(CLIENT_ID, &config).await?);
+    let filter = TopicFilter::new(SOURCE_TOPIC)?.qos(QoS::AtMostOnce);
+
+    let mut errors = mqtt.subscribe_errors();
+    let mut messages: MessageStream = mqtt.subscribe(filter).await?;
+
+    let mut interval =
```
Comment (vaillant) [inline dm-agent/src/main.rs:52]: Is that correct here? It feels that the statement to start the 1s interval should be started when the first message is received (it was specified in the doc like that).
However, does this code just raises "interval.tick()" every 1s, starting at a more-or-less ramdom time?

File: dm-agent/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+            error = errors.next().fuse() => {
+                if let Some(err) = error {
+                    error!("{}", err);
+                }
+            }
+
+            maybe_message = messages.next().fuse() => {
+                if let Some(message) = maybe_message {
+                    if message_batch.is_empty() {
+                        let _ = message_batch.timestamp(&current_timestamp()).map_err(|err| error!("{}", err));
```
Comment (vaillant) [inline dm-agent/src/main.rs:68]: At this point I would expect to see that the 1s intervall is started. 
See above message.

File: dm-agent/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_message_batch_processor() {
+        let mut message_batch = MeasurementGrouper::new();
+
+        let topic = Topic::new("collectd/temperature/value").unwrap();
```
Comment (vaillant) [inline dm-agent/src/main.rs:139]: Collectd topic names include also an instance (computer), that is not here in this test data.

See from study:
collectd/alsu-VirtualBox/users/users 1616147837.755:1
collectd/alsu-VirtualBox/swap/swap_io-out 1616147837.754:0
collectd/alsu-VirtualBox/swap/swap_io-in 1616147837.754:0
collectd/alsu-VirtualBox/memory/memory-used 1616147837.758:1085927424
collectd/alsu-VirtualBox/memory/memory-cached 1616147837.758:1374941184

File: dm-agent/src/main.rs [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+    }
+}
+
+fn tedge_json_from_collectd_message(
+    message_batch: &mut MeasurementGrouper,
+    message: Message,
+) -> anyhow::Result<()> {
+    let topic = message.topic.name.clone();
+    let split: Vec<&str> = topic.split('/').collect();
+    let group_measurement_key = split[split.len() - 2];
```
Comment (vaillant) [inline dm-agent/src/main.rs:90]: I would expect to see that there is a check if the topic is really from collectd, i.e. if someone just does
`tedge mqtt pub collectd/test "jaosda"` 
the process should not stop.

e.g.
`if split.len() < 4 { raise error }`

File: dm-agent/src/main.rs [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+    let measurement_key = split[split.len() - 1];
+
+    let payload = message.payload;
+    let payload = from_utf8(&payload)?;
+    let split: Vec<&str> = payload.split(':').collect();
+
+    let measurement_value = split[1].trim_end_matches(char::from(0)).parse::<f64>()?;
```
Comment (vaillant) [inline dm-agent/src/main.rs:97]: Again, check that there is at least one ":" in the payload, otherwise raise error.

Comment (vaillant) [inline dm-agent/src/main.rs:97]: (minor comment) I did not understood  what is the purpose of char::from (0). Somehow I was not even able to find the rust doc for that....

Comment (vaillant) [inline dm-agent/src/main.rs:97]: The value can be an int, and treating int's as float is always dangerous. Have you thought about this? 
At minimum, please add a test case with a value of MAX_INT and MIN_INT and check it works.

File: dm-agent/src/main.rs [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+    mqtt.as_ref().publish(tedge_message).await?;
+
+    Ok(())
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
```
Comment (vaillant) [inline dm-agent/src/main.rs:135]: There is no negative test case, there should be some added (wrong topic, wrong payload, etc.)

**Review by vaillant — COMMENTED** [review]
I left quite some comments.
Summary: 
* For external input to a process "serious" input data checks are needed.
* The intervall timing seems not to start on first message (but I might be wrong).

File: dm-agent/src/main.rs [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+            70.0
+        );
+        assert_eq!(
+            message_batch
+                .get_measurement_value(Some("coordinate"), "z")
+                .unwrap(),
+            90.0
+        );
+    }
+}
```
Comment (makr11st) [inline dm-agent/src/main.rs:190]: Could we split that test into separate tests and name them properly, eg:

```rust
fn message_batch_processor_temparature_value_correct() {
        let topic = Topic::new("collectd/temperature/value").unwrap();
        let collectd_message = Message::new(&topic, "123456789:32.5");
        tedge_json_from_collectd_message(&mut message_batch, collectd_message).unwrap();

        assert_eq!(
            message_batch
                .get_measurement_value(Some("temperature"), "value")
                .unwrap(),
            32.5
        );
}

**Review by makr11st — COMMENTED** [review]
The main of this binary should be split into lib where you implement the functionality and binary file which uses that lib to execute.
I think there are some problems especially with the interval not starting before data collection starting.

File: dm-agent/src/main.rs (hunk 1 above)
Comment (mneumann) [reply dm-agent/src/main.rs:52]: @vaillant your observation is correct. this is mainly to keep the code as simple as possible (for now).

File: dm-agent/src/main.rs [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+
+            _result = interval.tick() => {
+                if !message_batch.is_empty() {
+                    handle_message_batch(mqtt.clone(), message_batch);
+                    message_batch = MeasurementGrouper::new();
+                }
+            }
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (mneumann) [reply dm-agent/src/main.rs:82]: @makr11st note that the `hande_message_batch` will spawn a separate task for each batch.

File: dm-agent/src/main.rs (hunk 5 above)
Comment (albinsuresh) [reply dm-agent/src/main.rs:97]: The messages coming from collectd are suffixed with the uncode null character `\u0` which needs to be stripped before we can use that payload.

File: dm-agent/src/main.rs (hunk 1 above)
Comment (vaillant) [reply dm-agent/src/main.rs:52]: OK, but then this code is not correct and should be changed.

File: dm-agent/src/main.rs (hunk 2 above)
Comment (didier-wenzek) [reply dm-agent/src/main.rs:68]: Agree with Stefan. The collection interval is the maximum time we wait for messages after a first one - not the interval between the collection starts. Here, one might cut a batch that just happens to start a few milliseconds before the timeout.

**Review by didier-wenzek — COMMENTED** [review]
Not much to add to the previous comments.

* The current parsing of the collectd messages will panics in various way if given unexpected inputs. These error cases must be unit-tested and properly handled with `Err`
* The duration of a collection should be computed from the first message of a batch.
* The logic must be extracted in a library. However this can be done in a follow-up task encompassing the c8y and azure mapper.

File: dm-agent/src/main.rs [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+
+fn tedge_json_from_collectd_message(
+    message_batch: &mut MeasurementGrouper,
+    message: Message,
+) -> anyhow::Result<()> {
+    let topic = message.topic.name.clone();
+    let split: Vec<&str> = topic.split('/').collect();
+    let group_measurement_key = split[split.len() - 2];
+    let measurement_key = split[split.len() - 1];
```
[the remark this answers is not in the document]
Comment (albinsuresh) [reply dm-agent/src/main.rs:91]: The scheme is as follows: `collectd/hostname/<plugin-name>/<metric-key>` always with 4 levels in the topic.

File: dm-agent/src/main.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (albinsuresh) [reply dm-agent/src/main.rs:97]: Not always integers. Here is a sample CPU measurement:
```
collectd/alsu-VirtualBox/cpu-0/cpu-idle 1619679584.354:99.2059121722705
```

File: mapper/thin_edge_json/src/group.rs [hunk 10]
```diff
@@ -187,6 +214,45 @@ mod tests {
...
+    fn get_measurement_value() -> anyhow::Result<()> {
+        let mut grouper = MeasurementGrouper::new();
+        grouper.measurement(None, "temperature", 32.5)?;
+        grouper.measurement(Some("coordinate"), "x", 50.0)?;
+        grouper.measurement(Some("coordinate"), "y", 70.0)?;
+        grouper.measurement(Some("coordinate"), "z", 90.0)?;
+        grouper.measurement(None, "pressure", 98.2)?;
+
+        assert_eq!(
+            grouper.get_measurement_value(None, "temperature").unwrap(),
```
[the remark this answers is not in the document]
Comment (albinsuresh) [reply mapper/thin_edge_json/src/group.rs:227]: When I'm using `assert_matches` I'm getting this warning: `floating-point types cannot be used in patterns`

File: dm-agent/src/main.rs (hunk 7 above)
Comment (albinsuresh) [reply dm-agent/src/main.rs:190]: Why do you think this needs splitting? The intent of the test is to see if the API under test can build a Thin Edge JSON message batch from multiple collectd messages, a mix of single valued and multi-valued measurements. Splitting will defeat that purpose. I could add more tests to see the handling of a batch with just one message. But not sure if that adds any value in the presence of this test.

**Review by PradeepKiruvale — COMMENTED** [review]
I have a question regarding the design of the  DM agent. If I remember correctly, you were mentioning that this DM agent will be a multi-threaded one. How these threads are spawned and for which task?. Since this agent runs on resource constraint devices and when there are not many cores/CPUs having too many threads is not a problem?.

File: dm-agent/src/main.rs (hunk 5 above)
Comment (albinsuresh) [reply dm-agent/src/main.rs:97]: The conversions will work fine, even for `u128:MAX` and `i128:MIN` values, without any errors or panics, but with some precision loss as expected. Here are the results:
```
u128: 340282366920938463463374607431768211455 -> f64: 340282366920938500000000000000000000000.0
i128: -170141183460469231731687303715884105728 -> f64: -170141183460469230000000000000000000000.0
```
Have added both tests as well.

File: dm-agent/src/main.rs [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+            TIME_FORMAT.into(),
+        ))
+        .with_env_filter(filter)
+        .with_span_events(tracing_subscriber::fmt::format::FmtSpan::CLOSE)
+        .init();
+
+    info!("{} starting!", APP_NAME);
+
+    if let Err(err) = run().instrument(debug_span!(APP_NAME)).await {
+        error!("{}", err);
```
[the remark this answers is not in the document]
Comment (albinsuresh) [reply dm-agent/src/main.rs:39]: What I really wanted was to avoid this dm-agent from crashing even if it receives an invalid message. But now I realise that this isn't enough. What I need to do is to prevent the message processing select loop from crashing, but allow other unrecoverable errors to propagate up.

**Comment by albinsuresh** [comment]
> I have a question regarding the design of the DM agent. If I remember correctly, you were mentioning that this DM agent will be a multi-threaded one. How these threads are spawned?, I mean per topic one thread?. Since this agent runs on resource constraint devices and when there are not many cores/CPUs having too many threads is not a problem?.

Not per message. An async task is spawned per message batch to serialize that message into Thin Edge JSON and publish it to mosquitto. The `handle_message_batch` method is doing that.

File: dm-agent/src/main.rs [hunk 12]
```diff
@@ -0,0 +1,354 @@
...
+        )?;
+        Ok(())
+    }
+
+    async fn flush_as_mqtt_message(&self, mqtt: &Client) -> Result<(), Error> {
+        if !self.is_empty() {
+            let mut tedge_json_serializer = ThinEdgeJsonSerializer::new();
+            self.message_grouper.accept(&mut tedge_json_serializer)?;
+
+            let topic = Topic::new(TARGET_TOPIC).unwrap();
```
Comment (didier-wenzek) [inline dm-agent/src/main.rs:152]: Here is the kind of error that must be managed before the loop starts.

File: dm-agent/src/main.rs [hunk 13]
```diff
@@ -0,0 +1,354 @@
...
+        }
+    }
+}
+
+#[derive(Debug)]
+struct ThinEdgeMessage {
+    message_grouper: MeasurementGrouper,
+}
+
+impl ThinEdgeMessage {
```
Comment (didier-wenzek) [inline dm-agent/src/main.rs:112]: I have a mix-feeling with this struct which responsibility is not clear.

* A facade over the measurement grouper?
* A manager of the measurement collection (as suggested by the fact this struct assigns a timestamp)?
* A manager of the output over MQTT ?

File: dm-agent/src/main.rs [hunk 14]
```diff
@@ -0,0 +1,354 @@
...
+        self.message_grouper
+            .get_measurement_value(group_key, measurement_key)
+    }
+
+    pub fn add_measurement(&mut self, collectd_message: CollectdMessage) -> Result<(), Error> {
+        if self.message_grouper.is_empty() {
+            let _ = self
+                .message_grouper
+                .timestamp(&current_timestamp())
+                .map_err(|err| error!("{}", err));
```
Comment (didier-wenzek) [inline dm-agent/src/main.rs:137]: The fact that the timestamp is assigned internally will make unclear the relationship between this timestamp and the collection time window.

File: dm-agent/src/main.rs [hunk 15]
```diff
@@ -0,0 +1,354 @@
...
+
+    #[error(
+        "Message received on invalid collectd topic: {0}. \
+        Collectd message topics must be in the format collectd/<hostname>/<metric-plugin-name>/<metric-key>"
+    )]
+    InvalidMeasurementTopic(String),
+
+    #[error("Non UTF-8 payload received on topic: {0}")]
+    InvalidMeasurementPayload(String),
+
```
Comment (didier-wenzek) [inline dm-agent/src/main.rs:44]: Are all these errors required?

* It looks like as some copy/paste from other parts of the code.
* I would expect a single `#[error(transparent)]  ThinEdgeJsonSerializationError `.
* A refactoring of the thin-edge-json parser error type might be required.

**Review by didier-wenzek — COMMENTED** [review]
Despite some real improvements, this PR still deserves quite a lot of work.

* Move the collectd specific code under a module or library.
* Clarify the responsibility of the misc structs and methods.
* Fix the collect time issue.
* Fix the misc TODO.

File: dm-agent/src/main.rs (hunk 15 above)
Comment (albinsuresh) [reply dm-agent/src/main.rs:44]: These are collectd payload related errors not Thin Edge JSON payload related. For all the Thin Edge JSON errors there's a single error type already defined, called `ThinEdgeJsonSerializationError` exactly as you've suggested.

File: dm-agent/src/main.rs (hunk 14 above)
Comment (albinsuresh) [reply dm-agent/src/main.rs:137]: Not sure if I understood that suggestion completely. The timestamp is generated on the receipt of the very first message and NOT when the collection window starts. It was done this way due to the limitation around the collection window trigger in the current implementation which can lead to the window start time and first message receipt time to be different. Is that what you were highlighting? In that case, once we fix the collection trigger issue, this won't be an issue anymore.

File: dm-agent/src/main.rs (hunk 13 above)
Comment (albinsuresh) [reply dm-agent/src/main.rs:112]: This struct was introduced to cover the first two points. The third one could be moved out.

File: dm-agent/Cargo.toml [hunk 16]
```diff
@@ -0,0 +1,21 @@
+[package]
+name = "tedge_dm_agent"
```
[the remark this answers is not in the document]
Comment (albinsuresh) [reply dm-agent/Cargo.toml:2]: Will do this right before the merge.

File: dm-agent/src/mqtt.rs [hunk 17]
```diff
@@ -0,0 +1,76 @@
+use std::sync::Arc;
```
Comment (albinsuresh) [inline dm-agent/src/mqtt.rs:1]: All the traits in this module were created to ease unit testing the MQTT interactions from this dm-agent. This whole module can be moved to our `mqtt_client` library code if there's interest.

File: dm-agent/src/batcher.rs [hunk 18]
```diff
@@ -0,0 +1,415 @@
...
+        while let Some(message) = messages.next().await {
+            //Start a new message batch
+            message_batch.start_batch(message)?;
+
+            //Add messages to the current batch until the batching window times out and return
+            self.build_message_batch_with_timeout(&mut message_batch, messages, batching_window)
+                .await?;
+
+            //Close the current batch and send it to the batch processor
+            if let Some(batch) = message_batch.end_batch() {
```
Comment (didier-wenzek) [inline dm-agent/src/batcher.rs:154]: One will never see an empty batch here, since we added a message line 147.

This is not a big issue, but this highlight that the `message_batch` should be created & populated by the `build_message_batch_with_timeout` method.

**Review by didier-wenzek — COMMENTED** [review]
Overall the code is really good. However, I have a couple of questions and comments.

* Why not simply send the thin-edge-json messages from the main batcher loop? I see no obvious benefits to have two threads and an intermediate channels. Is this done because the publish method is async while the channel send is not? Does the added complexity worth the value? If so why not encapsulate this into the new `MqttClientImpl` struct?
* I believe that the `MessageBatcher` struct is of little benefits and that the `replace` function call can be avoided simply by not trying to re-use the `MessageBatcher` in the loop (by the way, this reuse provides no benefits since the *whole* object is replaced on each cycle of the loop).
* The `run`, `run_message_batcher` and `build_message_batch_with_timeout` functions can be largely simplified and improved by creating the batch inside the latter.

File: dm-agent/src/batcher.rs [hunk 19]
```diff
@@ -0,0 +1,415 @@
...
+                self.sender.send(batch)?;
```
[the remark this answers is not in the document]
Comment (didier-wenzek) [reply dm-agent/src/batcher.rs]: I think that is useless to have two threads here - one reading the input and the other dealing with the output. Indeed, by design we send a batch *only* when we have no more incoming messages.

**Review by didier-wenzek — APPROVED** [review]
Approved.

Some points can be improved as follow-up tasks.
* The new `MqttClient` trait introduced to ease testing can be moved in the mqtt_client crate and be used to test `tedge mqtt`.
* The logic of an independent thread publishing messages asynchronously can be moved into the `MqttClientImpl ` struct.

File: mapper/thin_edge_json/src/group.rs [hunk 20]
```diff
@@ -36,6 +36,27 @@ impl MeasurementGrouper {
         self.values.is_empty()
     }
 
+    pub fn clear(&mut self) {
+        self.values.clear();
+    }
```
[the remark this answers is not in the document]
Comment (albinsuresh) [reply mapper/thin_edge_json/src/group.rs:41]: Not anymore. Will remove it.
